### PR TITLE
Adding optional eagerClose parameter to merge methods in FlowOps and javadsl method overloads

### PR DESCRIPTION
**Pull request by Athelionas**
#18948

### CODE REVIEW

**Comment by drewhk** [comment]
OK TO TEST

**Comment by Athelionas** [comment]
It looks like the build has failed due to #18961 having run after this build

**Comment by ktoso** [comment]
PLS BUILD

**Comment by ktoso** [comment]
LGTM, thanks @Athelionas!

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 1]
```diff
@@ -929,7 +929,22 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
    * '''Cancels when''' downstream cancels
    */
   def merge[T >: Out](that: Graph[SourceShape[T], _]): javadsl.Flow[In, T, Mat] =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Doesn't document that it uses false eager close.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 2]
```diff
@@ -929,7 +929,22 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
...
+   *
+   * '''Emits when''' one of the inputs has an element available
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' all upstreams complete (eagerClose=false) or one upstream completes (eagerClose=true)
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def merge[T >: Out](that: Graph[SourceShape[T], _], eagerClose: Boolean): javadsl.Flow[In, T, Mat] =
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Doesn't document what eager close is.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 3]
```diff
@@ -939,6 +954,17 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
    */
   def mergeMat[T >: Out, M, M2](that: Graph[SourceShape[T], M],
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Doesn't document that it uses false eager close.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 4]
```diff
@@ -939,6 +954,17 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
...
                                 matF: function.Function2[Mat, M, M2]): javadsl.Flow[In, T, M2] =
+    mergeMat(that, matF, eagerClose = false)
+
+  /**
+   * Merge the given [[Source]] to this [[Flow]], taking elements as they arrive from input streams,
+   * picking randomly when several elements ready.
+   *
+   * @see [[#merge]]
+   */
+  def mergeMat[T >: Out, M, M2](that: Graph[SourceShape[T], M],
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Doesn't document what eager close is.

**Comment by viktorklang** [comment]
Hi!

I commented that there was a bit of missing documentation,

Since streams do not close, but complete, I'd propose that it is `eagerComplete` rather than `eagerClose`. What do you think of that?

**Comment by Athelionas** [comment]
Hi!

Thanks for the review, I'll add missing pieces to the documentation.

I named the parameter after the eagerClose param of Merge and MergePreferred, but I agree. IMHO, it should have been called eagerComplete there too. If we decide to refactor the parameter names then we should do it consistently everywhere.

**Comment by viktorklang** [comment]
@Athelionas I completely agree. And if we rename them on Merge and MergePreferred we should add a section to the migration guides (for those who might have used their current name in code)

**Comment by ktoso** [comment]
In theory we could use `@deprecatedName` on the params to ease migration, not its worth it during Milestone phase though

**Comment by drewhk** [comment]
+1 for rename

**Comment by Athelionas** [comment]
I checked the documentation and I don't really know what should be added.

For example this line makes it quite clear what eagerClose (or eagerComplete) is:
`'''Completes when''' all upstreams complete (eagerClose=false) or one upstream completes (eagerClose=true)`

 Default values should be stated though.

In case of javadsl merge method without eagerClose parameter, `'''Completes when''' all upstreams complete` explains the behaviour quite well. Notes about parameters passed to underlying methods seem unnecessary implementation details to me.

What's your opinion?

**Comment by drewhk** [comment]
@Athelionas There were two changes discussed:

 - use eagerComplete instead of eagerClose in *all* stream stages, not just this use. (TCP is the only exception where close is, well, close).
 - document this name change in the migration guide (`akka\akka-docs-dev\rst\scala\migration-guide-1.0-2.x-scala.rst` and `akka\akka-docs-dev\rst\java\migration-guide-1.0-2.x-java.rst`)

**Comment by Athelionas** [comment]
First point is done hopefully, second is in progress :)

**Comment by ktoso** [comment]
Hi @Athelionas,
just a friendly ping if you're still working on the remaining comments to this PR? :-)
It's a nice improvement and we'd like to pull it in :-)

Let us know if you're stuck with some part of it.

File: akka-docs-dev/rst/java/migration-guide-1.0-2.x-java.rst [hunk 5]
```diff
@@ -8,7 +8,7 @@ The 2.0 release contains some structural changes that require some
 simple, mechanical source-level changes in client code.
 
 
-Introduced proper named constructor methods insted of ``wrap()``
+Introduced proper named constructor methods instead of ``wrap()``
 ================================================================
```
Comment (rkuhn) [inline akka-docs-dev/rst/java/migration-guide-1.0-2.x-java.rst]: please fix heading marker as well (one `=` is missing now, sphinx is quite picky … )

File: akka-docs-dev/rst/scala/migration-guide-1.0-2.x-scala.rst [hunk 6]
```diff
@@ -8,7 +8,7 @@ The 2.0 release contains some structural changes that require some
 simple, mechanical source-level changes in client code.
 
 
-Introduced proper named constructor methods insted of ``wrap()``
+Introduced proper named constructor methods instead of ``wrap()``
 ================================================================
```
Comment (rkuhn) [inline akka-docs-dev/rst/scala/migration-guide-1.0-2.x-scala.rst]: here as well

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 1 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: well, it does: the «completes when» clause clearly describes the behavior, that’s good enough for me

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: if `@see` works then that will be good enough

**Comment by rkuhn** [comment]
@Athelionas please let us know if you intend to pick this up again, otherwise we’ll take it over on Monday. In any case thanks for the contribution!

**Comment by ktoso** [comment]
OK TO TEST

**Comment by rkuhn** [comment]
Taking over.

**Pull request closed without merging**